## Background

#### The tna operon

The tryptophanase (<i>tna<\i> operon of E. coli and other bacteria controls the influx and breakdown of externally-available tryptophan. In the absence of tryptophan, rho associates with the transcript near the stop codon of tnaC and stops transcription before the polymerase reaches the downstream gene. However, in the presence of high concentrations of externally-available tryptophan, the ribosome stalls at the stop codon of tnaC and occludes rho from associating with the transcript, allowing expression of the downstream genes.

![plot](./images/Gong2002.png)
The tnaC mechanism

#### Assay design
The objective of this study was to sort members of a variant library for their ability to stall the ribosome using a reporter assay in the presence/absence of tryptophan. A plasmid library of over 3,250 bioinformatically-identified variants was derived by searching over 10,000 bacterial genomes using the motif search algorithm. Genscript prepared a plasmid library to incorporate the variants across a 60mer stretch of nucleotides within the tnaC gene.  This regulatory sequence was placed upstream of GFP to induce fluorescence upon ribosome stalling. All of the variants were designed to begin with MNIL (encoded by ATGAATATCCTG) followed by 57 variable nucleotides and a stop codon (TGA or TAA).  
 
![plot](./images/tnaC_schematic.jpeg)
Modifications made to the <i>tna<\i> operon for the assessment of the variant library

The library was cloned into E. coli and those transformed with the library were selected with kanamycin. The transformed pool was grown to OD~0.7 in LB, starved for tryptophan via a 1:10 dilution into casamino acid, and then fed tryptophan.  The difference in reporter fluorescence was identified by a cell sorter to enrich members that allowed ribosome stalling. Following grow out, plasmids were isolated from the culture and an NGS library was prepared by amplification from the constant regions on the plasmids.  These were then multiplexed and run on an an Illumina MiSeq.

The reference used in this study was generated by Genscript during their NGS verifiction of the original library (version 3, generated around 5/19/2019).  The 3 libraries were derived from this original via selection methods. This script aims to observe how the distribution of membership changes from the original population to the 3 populations.  The purpose is to demonstrate an approach that could be used for directed evolution or pooled screening efforts.

The libraries had been pre-processed in 2020 and the semi-processed data files are those available locally.

| Library | Method | Start_count | Prot_count |
| --- | --- | --- | --- |
| Reference | Already trimmed (20aas) | 3248 | 3248 |
| lib_1 | Trimmed but doesn't have MNIL as first 4 aas (72nts ~24 aas) | 59529 | 59529
| lib_2 | Split with GTACC 24 aas (should probably do this with an anchor, and multiple expressions) | 51218 | 860 (25% 'M') |
| lib_3 | Was already split at MNIL (20 aas) | 29804 |29804| 29767|


libary files use either AGGTACC as an anchor (generates a 73 nt variable sequence) or start at the end of the MNIL constant region.






Once seuqencing data was found to be much more highly variable than the input library of 3,200 sequences, I took two approaches to explore enrichment. The first approach uses peptide frequencies under the assumption that any mutation may generate more diversity that would be important at the peptide level, since the peptide is what is understood to interact with the free tryptophan within the ribosome exit tunnel.  

Later in the script, variants are mapped back to their closest original library member (via Levenshtein distance) in the event that the diversity just came from library prep/sequencing artifacts.


# Calculate relative proportion of membership for terminal peptides
# The terminal region of tnaC is responsible for stalling the ribosome, 
# so if cloning did not go as expected, the terminal region could still be 
# responsible for signal


#one hot encode peptide sequences, pad sequences with "-" if needed

### The script below is used to investigate individual libraries
The script is designed for stand alone outputs, and not for use in a python notebook; however, users can be helpful to explore the seqeunces of interest within this pipeline.  The dataframes of individual libraries and columns should be customized to the library of interest.

In order to use this in interactive mode, a user would have to save the dataframe of interest and put the script into a standalone python script and then call the script by: 'bokeh serve --show appname.py'

### Below is to plot the full dataset against reference
Here, we save a csv of the data of interest so that the script can be run as standalone output, which will enable interactive selection and CSV generation from selected points.

Norm_freq1_corr=ref_123_df['Norm_freq1']-ref_123_df['Norm_freq'],





### Nucleotide-based analysis
There appear to be satellite clusters around a major point when looking at individual datasets, which suggests to me that there were a lot of mutations during library prep, evolution, sequencing, etc.  Even points just one amino acid away from the major point is almost negligable when it comes to frequency.  This effect may be real and changing one amino acid from the unique sequence could stop functional activity; but it's likely an artifact of sequencing, as not every position is likely to be critical to function.  Therefore, I'll attempt to use a distance measure to assign each variant to it's closest reference sequence at the nucleotide level. 


| Model | Description | Substitutions | Indels |
| --- | --- | --- | --- |
| Hamming distance | measures the number of substitutions from one string to another. Requires fixed length| X ||
|Levenshtein distance | measures edit distance between strings (Best for this use case because we expect artifact mutations) | X | X | 
|Smith-Waterman algorithm| identifies regions of similarity between two sequences | X | X |
|Needleman-Wunsch algorithm| global alignment to compare two sequences and identify regions of similarity| X | X | X |
|Blosum Matrix| Substitution matrix that assigns scores to each possible amino acid substitution based on the frequency of occurrence of the substitution in known protein sequences | X|X|
|Phylogenetic analysis| Compares multiple amino acid sequences via construction of a phylogenetic tree and measures branch lengths ||





Now, standalone apps can be run for the peptide sequences or the clustered nucleotide sequences. Variants of interest can be selected using the lasso tool and extracted to a CSV.

1. Using the terminal, activate the virtual environment with bokeh \
2. Set the library to the one you want to view \
3. Enter the command from the main folder: 
    - bokeh serve --show ./code/nuc_app.py 
    - bokeh serve --show ./code/pep_app.py
